stlab/concurrency/router.hpp

```diff
@@ -0,0 +1,137 @@
+/*
+    Copyright 2015 Adobe
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+
+/**************************************************************************************************/
+
+#ifndef STLAB_CONCURRENCY_ROUTER_HPP
+#define STLAB_CONCURRENCY_ROUTER_HPP
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include <stlab/concurrency/channel.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+inline namespace v1 {
+
+/**************************************************************************************************/
+
+template <class T, class K>
+class router {
+    using channel_pair = std::pair<stlab::sender<T>, stlab::receiver<T>>;
+    using route_pair = std::pair<K, channel_pair>;
+    using routes = std::vector<route_pair>;
+
+public:
+    template <class E, class F>
+    router(E executor, F router_func) : _self{std::make_shared<model<E, F>>(std::move(executor), std::move(router_func))} {}
```

Comment:
Is there a reason that _self is a shared_ptr? Is it shared between others? If not then it should be a unique_ptr. See Sean Parents talk: https://sean-parent.stlab.cc/papers-and-presentations/#inheritance-is-the-base-class-of-evil

What about the other functions that make this a regular type?

Reply:
`std::unique_ptr<>` can be used only if the following is valid:
```c++
router.set_ready();
auto hold = receive | std::bind(std::move(router), _1);
```
Is it valid to `set_ready()` before completing the graph?

Reply:
Never mind, that subgraph is dependent on the `receive.set_ready()` not `router.set_ready()`

Reply:
I guess the only reason you would really want `std::shared_ptr<>` is if you plan on having multiple entry points for the router and not just one... perhaps it is better to use `std::shared_ptr<>`. Thoughts?